File: server/internal/model/task_activity.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

type TaskActivityAction string

const (
	ActionTaskCreated        TaskActivityAction = "task_created"
	ActionTitleUpdated       TaskActivityAction = "title_updated"
	ActionDescriptionUpdated TaskActivityAction = "description_updated"
	ActionStatusUpdated      TaskActivityAction = "status_updated"
	ActionDueDateUpdated     TaskActivityAction = "due_date_updated"
	ActionProjectChanged     TaskActivityAction = "project_changed"
	ActionAttachmentsUpdated TaskActivityAction = "attachments_updated"
	ActionTagAdded           TaskActivityAction = "tag_added"
	ActionTagRemoved         TaskActivityAction = "tag_removed"
	ActionMovedToTrash       TaskActivityAction = "task_moved_to_trash"
	ActionRestored           TaskActivityAction = "task_restored"
	ActionAbandoned          TaskActivityAction = "task_abandoned"
	ActionReactivated        TaskActivityAction = "task_reactivated"
)

type TaskActivity struct {
	ID        uuid.UUID          `gorm:"type:uuid;primaryKey" json:"id"`
	TaskID    uuid.UUID          `gorm:"type:uuid;index;not null" json:"task_id"`
	UserID    *uuid.UUID         `gorm:"type:uuid;index" json:"user_id"`
	Action    TaskActivityAction `gorm:"size:50;not null" json:"action"`
	Metadata  datatypes.JSON     `gorm:"type:jsonb;default:'{}'" json:"metadata"`
	CreatedAt time.Time          `json:"created_at"`

	User *User `gorm:"foreignKey:UserID" json:"user,omitempty"`
}

func (a *TaskActivity) BeforeCreate(tx *gorm.DB) error {
	if a.ID == uuid.Nil {
		a.ID = uuid.New()
	}
	return nil
}

File: web/src/types/taskActivity.ts
export type TaskActivityAction =
  | "task_created"
  | "title_updated"
  | "description_updated"
  | "status_updated"
  | "due_date_updated"
  | "project_changed"
  | "attachments_updated"
  | "tag_added"
  | "tag_removed"
  | "task_moved_to_trash"
  | "task_restored"
  | "task_abandoned"
  | "task_reactivated"
  | string;

export interface TaskActivity {
  id: string;
  task_id: string;
  user_id?: string | null;
  anonymous_id?: string | null;
  action: TaskActivityAction;
  metadata: Record<string, unknown> | null;
  created_at: string;
}

export interface TaskActivityInput {
  action: TaskActivityAction;
  metadata?: Record<string, unknown>;
}
